src/libxtp/eigencuda.cc

```diff
@@ -0,0 +1,197 @@
+/*
+ *            Copyright 2009-2019 The VOTCA Development Team
+ *                       (http://www.votca.org)
+ *
+ *      Licensed under the Apache License, Version 2.0 (the "License")
+ *
+ * You may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *              http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#include <votca/xtp/eigencuda.h>
+
+namespace votca {
+namespace xtp {
+
+template <typename T>
+EigenCuda<T>::~EigenCuda() {
+  cublasDestroy(_handle);
+  for (auto &p : _allocated) this->fun_free(p.second);
+}
+
+template <typename T>
+void EigenCuda<T>::fun_alloc(T **x, std::size_t n) const {
+  // Allocate memory in the device
+  (_pinned) ? cudaMallocHost(x, n) : cudaMalloc(x, n);
+}
+
+template <typename T>
+void EigenCuda<T>::fun_free(T *x) const {
```

Comment:
what do you mean by fun_free?

Reply:
I am replacing the `fun_` functions with more meaningful names like `gpu_alloc` and `gpu_free`. Those function basically allocated and deallocate memory in the device